integration_tests/src/main/python/join_test.py

```diff
@@ -1151,3 +1151,63 @@ def do_join(spark):
         conf={"spark.sql.autoBroadcastJoinThreshold": "-1",
               "spark.sql.sources.useV1SourceList": "",
               "spark.rapids.sql.input." + scan_name: False})
+
+@ignore_order(local=True)
+@pytest.mark.parametrize("is_left_host_shuffle", [False, True], ids=idfn)
+@pytest.mark.parametrize("is_right_host_shuffle", [False, True], ids=idfn)
+@pytest.mark.parametrize("is_left_smaller", [False, True], ids=idfn)
+@pytest.mark.parametrize("batch_size", ["1024", "1g"], ids=idfn)
+def test_new_inner_join(is_left_host_shuffle, is_right_host_shuffle, is_left_smaller, batch_size):
+    join_conf = {
+        "spark.rapids.sql.join.useShuffledSymmetricHashJoin": "true",
+        "spark.sql.autoBroadcastJoinThreshold": "1",
+        "spark.rapids.sql.batchSizeBytes": batch_size
+    }
+    left_size, right_size = (2048, 1024) if is_left_smaller else (1024, 2048)
+    def do_join(spark):
+        left_df = gen_df(spark, [
+            ("key1", RepeatSeqGen(long_gen, length = 10)),
+            ("ints", int_gen),
+            ("key2", RepeatSeqGen(int_gen, length = 20)),
+            ("floats", float_gen)], left_size)
+        right_df = gen_df(spark, [
+            ("doubles", double_gen),
+            ("key2", RepeatSeqGen(int_gen, length = 30)),
```

Comment:
There is a lot of overlap here because we're using the exact same generator for the RepeatSeqGen.  The first 10 values that come out of `int_gen` for `key2` in one table will be the first 10 values that come out of `int_gen` for `key2` in the other table, but it will contine to generate values beyond the first 10.

I'll just switch this to explicit values, as that also addresses the "what if too many nulls" question above.

Reply:
That is not how it works. I put it a debug_df for the answer to the join and I got back 

```
COLLECTED
[]
```

RepeatSeqGen is using a random number generator that was passed into all of the columns, and then used in turn by each of them to generate the data. That means even if I have Key1 followed by Key2 in the same order in both datasets, if the lengths are different the second key will have a totally different seed that it starts out with compared to the second dataset. 

```
r = random(0)
(0 until 10) r.randint(longmin, longmax) #key1
(0 until 20) r.randint(intmin, intmax) #key2
```
vs
```
r = random(0)
(0 until 30) r.randint(intmin, intmax) #key2
(0 until 15) r.randint(longmin, longmax) #key1
```

Those are not going to produce overlapping key sets.

Reply:
Ah, so the bug there is using `int_gen` instead of `IntegerGen()` since it's reusing the same instance for both.

Reply:
Not really.

https://github.com/NVIDIA/spark-rapids/blob/b3839e49d0a71f0b48b979a663beab26552d1859/integration_tests/src/main/python/data_gen.py#L811-L815

Creates a single Random instance for generating the entire data frame. That random instance is passed to start, in this case for a Struct

https://github.com/NVIDIA/spark-rapids/blob/b3839e49d0a71f0b48b979a663beab26552d1859/integration_tests/src/main/python/data_gen.py#L531-L537

which will call start on all of that structs children in order.

If any of those children are RepSeqGen their start is special in that it calls start on its child, and then proceeds to use the rand passed in to generate the number of values requested.

https://github.com/NVIDIA/spark-rapids/blob/b3839e49d0a71f0b48b979a663beab26552d1859/integration_tests/src/main/python/data_gen.py#L372-L379

So if the number of unique values or the order of the columns change, then the range of values generated by RepSeqGen can are also likely to change.

It is by design that the random is shared between all of the items in a dataframe because if I generate int a, int b I don't want a and b to be the same, and this was a simple way of doing it.  There are downsides. If I add or remove a column all of the data after the first row is likely to change, as the data is generated row by row, not column by column. But that also means that if I add or remove rows, but keep the schema the same, then the data is likely to remain the same.

There should be no oddness using int_gen vs IntegerGen, as all of the state is stored in Random, and not in the Generator itself, outside of RepSeqGen.

Reply:
So it looks like test_broadcast_join_mixed has the same issues.  https://github.com/NVIDIA/spark-rapids/blob/branch-24.02/integration_tests/src/main/python/join_test.py#L632-L646

Reply:
Yes it looks like there is a problem with that one too.

Reply:
https://github.com/NVIDIA/spark-rapids/issues/10242